util/tracecpu/profile.go

```diff
@@ -0,0 +1,449 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tracecpu
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"runtime/pprof"
+	"strconv"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/google/pprof/profile"
+	"github.com/pingcap/tidb/config"
+	"github.com/pingcap/tidb/sessionctx/variable"
+	"github.com/pingcap/tidb/util"
+	"github.com/pingcap/tidb/util/logutil"
+	"go.uber.org/zap"
+)
+
+const (
+	labelSQL        = "sql"
+	labelSQLDigest  = "sql_digest"
+	labelPlanDigest = "plan_digest"
+)
+
+// GlobalTopSQLCPUProfiler is the global SQL stats profiler.
+var GlobalTopSQLCPUProfiler = newTopSQLCPUProfiler()
+
+type topSQLCPUProfiler struct {
+	taskCh     chan *profileTask
+	cacheBufCh chan *profileTask
+
+	mu struct {
+		sync.Mutex
+		ept *exportProfileTask
+	}
+	collector atomic.Value
+}
+
+// newTopSQLCPUProfiler create a topSQLCPUProfiler.
+func newTopSQLCPUProfiler() *topSQLCPUProfiler {
+	return &topSQLCPUProfiler{
+		taskCh:     make(chan *profileTask, 128),
+		cacheBufCh: make(chan *profileTask, 128),
+	}
+}
+
+func (sp *topSQLCPUProfiler) Run() {
+	logutil.BgLogger().Info("cpu profiler started")
+	go sp.startCPUProfileWorker()
+	go sp.startAnalyzeProfileWorker()
+}
+
+func (sp *topSQLCPUProfiler) SetCollector(c TopSQLCollector) {
+	sp.collector.Store(c)
+}
+
+func (sp *topSQLCPUProfiler) getCollector() TopSQLCollector {
+	c, ok := sp.collector.Load().(TopSQLCollector)
+	if !ok || c == nil {
+		return nil
+	}
+	return c
+}
+
+func (sp *topSQLCPUProfiler) RegisterSQL(sqlDigest, normalizedSQL string) {
+	c := sp.getCollector()
+	if c == nil {
+		return
+	}
+	c.RegisterSQL(sqlDigest, normalizedSQL)
+}
+
+func (sp *topSQLCPUProfiler) RegisterPlan(planDigest string, normalizedPlan string) {
+	c := sp.getCollector()
+	if c == nil {
+		return
+	}
+	c.RegisterPlan(planDigest, normalizedPlan)
+}
+
+func (sp *topSQLCPUProfiler) startCPUProfileWorker() {
+	defer util.Recover("top-sql", "profileWorker", nil, false)
+	for {
+		if sp.IsEnabled() {
+			sp.doCPUProfile()
+		} else {
+			time.Sleep(time.Second)
+		}
+	}
+}
+
+func (sp *topSQLCPUProfiler) doCPUProfile() {
+	interval := config.GetGlobalConfig().TopSQL.RefreshInterval
+	task := sp.newProfileTask()
+	if err := pprof.StartCPUProfile(task.buf); err != nil {
+		// Sleep a while before retry.
+		time.Sleep(time.Millisecond)
+		sp.putTaskToBuffer(task)
+		return
+	}
+	ns := int(time.Second)*interval - time.Now().Nanosecond()
+	time.Sleep(time.Nanosecond * time.Duration(ns))
+	pprof.StopCPUProfile()
+	sp.sendProfileTask(task)
+}
+
+func (sp *topSQLCPUProfiler) sendProfileTask(task *profileTask) {
+	task.end = time.Now().Unix()
+	sp.taskCh <- task
+}
+
+func (sp *topSQLCPUProfiler) startAnalyzeProfileWorker() {
+	defer util.Recover("top-sql", "analyzeProfileWorker", nil, false)
+	for {
+		task := <-sp.taskCh
+		reader := bytes.NewReader(task.buf.Bytes())
+		p, err := profile.Parse(reader)
```

Comment:
Seems that you shoud use [`ParseData`](https://github.com/google/pprof/blob/master/profile/profile.go#L156), since inside `Parse` it simply [turns a Reader back to the buffer](https://github.com/google/pprof/blob/master/profile/profile.go#L147), and you already have a buffer.

Reply:
Nice catch.